# HG changeset patch
# User Roland Haas <rhaas@caltech.edu>
# Date 1317396741 25200
# Node ID 20b165111ef71c4f7484105a8d2ecd1158a95869
# Parent  f35e068d92845ba80ab506b2dbd04d805cbb1fa5
Carpet: Correct schedule and restriction traversal for postrestrictinitial

this is analogous to b4429f4006e5 and makes sure that all restrictions happen
before routines in POSTRESTRICTINITIAL are called (which might trigger
prolongation when doing a SYNC, thus pulling data from coarser levels.)

diff -r f35e068d9284 -r 20b165111ef7 Carpet/Carpet/src/Initialise.cc
--- a/Carpet/Carpet/src/Initialise.cc	Thu Sep 08 19:40:44 2011 -0400
+++ b/Carpet/Carpet/src/Initialise.cc	Fri Sep 30 08:32:21 2011 -0700
@@ -459,29 +459,65 @@
     static Timer timer ("CallRestrict");
     timer.start();
     
-    for (int rl=reflevels-2; rl>=0; --rl) {
-      BEGIN_MGLEVEL_LOOP(cctkGH) {
-        ENTER_LEVEL_MODE (cctkGH, rl) {
-          BeginTimingLevel (cctkGH);
+    for (int ml=mglevels-1; ml>=0; --ml) {
+
+      for (int rl=reflevels-2; rl>=0; --rl) {
+        ENTER_GLOBAL_MODE (cctkGH, ml) {
+          ENTER_LEVEL_MODE (cctkGH, rl) {
+            BeginTimingLevel (cctkGH);
                     
-          do_early_global_mode = reflevel==reflevels-2;
-          do_late_global_mode = reflevel==0;
-          do_early_meta_mode = do_early_global_mode and mglevel==mglevels-1;
-          do_late_meta_mode = do_late_global_mode and mglevel==0;
-          do_global_mode = do_late_global_mode; // on last iteration, finest grid
-          do_meta_mode = do_late_meta_mode; // on last iteration, finest grid
+            Waypoint ("Initialisation/Restrict at iteration %d time %g",
+                      cctkGH->cctk_iteration, (double)cctkGH->cctk_time);
+          
+            Restrict (cctkGH);
 
-          Waypoint ("Initialisation/Restrict at iteration %d time %g",
-                    cctkGH->cctk_iteration, (double)cctkGH->cctk_time);
+            EndTimingLevel (cctkGH);
+          } LEAVE_LEVEL_MODE;
+        } LEAVE_GLOBAL_MODE;
+      } // for rl
+
+      bool have_done_global_mode = false;
+      bool have_done_early_global_mode = false;
+      bool have_done_late_global_mode = false;
+      bool have_done_anything = false;
+        
+      for (int rl=0; rl<reflevels; ++rl) {
+        ENTER_GLOBAL_MODE (cctkGH, ml) {
+          ENTER_LEVEL_MODE (cctkGH, rl) {
+            BeginTimingLevel (cctkGH);
+                    
+            do_early_global_mode = not have_done_early_global_mode;
+            do_late_global_mode = reflevel==reflevels-1;
+            do_early_meta_mode =
+              do_early_global_mode and mglevel==mglevels-1;
+            do_late_meta_mode = do_late_global_mode and mglevel==0;
+            do_global_mode = do_late_global_mode;
+            do_meta_mode = do_global_mode and do_late_meta_mode;
+            assert (not (have_done_global_mode and do_global_mode));
+            assert (not (have_done_early_global_mode and
+                         do_early_global_mode));
+            assert (not (have_done_late_global_mode and
+                         do_late_global_mode));
+            have_done_global_mode |= do_global_mode;
+            have_done_early_global_mode |= do_early_global_mode;
+            have_done_late_global_mode |= do_late_global_mode;
+            have_done_anything = true;
+
+            Waypoint ("Initialisation/PostRestrict at iteration %d time %g",
+                      cctkGH->cctk_iteration, (double)cctkGH->cctk_time);
+                
+            ScheduleTraverse (where, "CCTK_POSTRESTRICTINITIAL", cctkGH);
           
-          Restrict (cctkGH);
-          
-          ScheduleTraverse (where, "CCTK_POSTRESTRICTINITIAL", cctkGH);
-          
-          EndTimingLevel (cctkGH);
-        } LEAVE_LEVEL_MODE;
-      } END_MGLEVEL_LOOP;
-    } // for rl
+            EndTimingLevel (cctkGH);
+          } LEAVE_LEVEL_MODE;
+        } LEAVE_GLOBAL_MODE;
+      } // for rl
+
+      if (have_done_anything) assert (have_done_global_mode);
+      if (have_done_anything) assert (have_done_early_global_mode);
+      if (have_done_anything) assert (have_done_late_global_mode);
+        
+    } // for ml
     
     timer.stop();
   }
